### PR TITLE
Kucoin: Changes default subscription channel for tickers and orderbooks

### PR DESCRIPTION
**Pull request by shazbert**
# PR Description

* ~~There was a lag process involving numerous outbound subscriptions, this helps alleviate that by batching calls [spot only]~~ This was done in a PR that was merged first and it was much better than this. 🚀 
* ~~This also brings subscription errors to the surface~~ Not needed anymore
* Changes default subscription channel for spot tickers `marketTickerChannel` to `marketAllTickersChannel` as this is counted as one subscription and allows more potential subscriptions to occur for orderbooks. 
* Changes default subscription channel for spot orderbooks `marketOrderbookLevel2Channels` to `marketOrderbookLevel2to5Channel` so as to not send subsequent REST requests to build books. This has a limitation of not being real time (ms latency not defined anyway) to 100ms. 
* optimise `listOfAssetsCurrencyPairEnabledFor` and refactor caller usage
* Fix handling of orderbook depth 5 && 20 orderbook snapshots (REST request not needed)
* Fixes issue where there is a deadlock when there is a slow subscriber when all pairs enabled. [211aa6f](https://github.com/thrasher-corp/gocryptotrader/pull/1371/commits/211aa6faa28e43579a0e4656ff56704f97701d75)

Notes:

* ~~This is only for SPOT~~
* There is a subscription limit per connection at ~300

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Nice code, no real complaints there. The ticker:all endpoint does seem pretty wasteful for the majority of usecases, but its all working well enough

**Review by gloriousCode — APPROVED** [review]
This is good for ~bit~kucoin. tACK!

Two extra notes: an unrelated test now fails here (`TestGetFuturesRealTimeTicker`) because its not on latest, but will be resolved on merge. This also may impact https://github.com/thrasher-corp/gocryptotrader/pull/1394

**Comment by shazbert** [comment]
I shall mergerino

**Review by thrasher- — CHANGES_REQUESTED** [review]
One minor nit

File: exchanges/kucoin/kucoin_websocket.go [hunk 1]
```diff
@@ -1678,13 +1724,20 @@ func (o *orderbookManager) stopNeedsFetchingBook(pair currency.Pair, assetType a
 	return nil
 }
 
-func (ku *Kucoin) listOfAssetsCurrencyPairEnabledFor(cp currency.Pair) []asset.Item {
-	assets := []asset.Item{}
-	for _, a := range ku.CurrencyPairs.GetAssetTypes(true) {
-		pairs, err := ku.GetEnabledPairs(a)
-		if err == nil && pairs.Contains(cp, true) {
-			assets = append(assets, a)
+func (ku *Kucoin) listOfAssetsCurrencyPairEnabledFor(cp currency.Pair) ([]asset.Item, error) {
```
Comment (gloriousCode) [inline exchanges/kucoin/kucoin_websocket.go:1727]: So..... While I appreciate your improvements here, I'd like to try and address this here if you're cool with that. If not, we can look into it for a future PR, that's okay. This is something I'm working on, I'll look at it with you tomorrow if you want as I expand testing for it.

I would like this function to more easily figure out the asset for the pair without calling such heavy functions.

Here's my code so far:
```go
func (ku *Kucoin) CalculateAssets(topic string, cp currency.Pair) ([]asset.Item, error) {
	if cp.Quote.Equal(currency.USDTM) || strings.Contains(topic, "contract") {
		if err := ku.CurrencyPairs.IsAssetEnabled(asset.Futures); err != nil {
			return nil, nil
		}
		return []asset.Item{asset.Futures}, nil
	}
	if strings.Contains(topic, "margin") {
		if err := ku.CurrencyPairs.IsAssetEnabled(asset.Margin); err != nil {
			return nil, nil
		}
		return []asset.Item{asset.Margin}, nil
	}
	cps := cp.String()
	resp := make([]asset.Item, 0, 2)
	_, spotEnabled, err := ku.MatchSymbolCheckEnabled(cps, asset.Spot, true)
	if err != nil && !errors.Is(currency.ErrCurrencyNotFound, err) {
		return nil, err
	}
	if spotEnabled {
		resp = append(resp, asset.Spot)
	}
	_, marginEnabled, err := ku.MatchSymbolCheckEnabled(cps, asset.Margin, true)
	if err != nil && !errors.Is(currency.ErrCurrencyNotFound, err) {
		return nil, err
	}
	if marginEnabled {
		resp = append(resp, asset.Margin)
	}
	return resp, nil
}


```

Here is my benchmarking:
```go
func BenchmarkButts(b *testing.B) {
	b.ReportAllocs()
	cases := []struct {
		pair           currency.Pair
		expectedResult []asset.Item
	}{
		{pair: spotTradablePair, expectedResult: []asset.Item{asset.Spot}},
		{pair: marginTradablePair, expectedResult: []asset.Item{asset.Spot, asset.Margin}},
		{pair: futuresTradablePair, expectedResult: []asset.Item{asset.Futures}},
	}
	for i := 0; i < b.N; i++ {
		for j := range cases {
			resp, err := ku.listOfAssetsCurrencyPairEnabledFor(cases[j].pair)
			if err != nil {
				b.Fatal(err)
			}
			for x := range resp {
				var found bool
				for y := range cases[j].expectedResult {
					if resp[x] == cases[j].expectedResult[y] {
						found = true
						break
					}
				}
				if !found {
					b.Error("bad match")
				}
			}
		}
	}
}

func BenchmarkButts2(b *testing.B) {
	b.ReportAllocs()
	cases := []struct {
		pair           currency.Pair
		topic          string
		expectedResult []asset.Item
	}{
		{pair: spotTradablePair, topic: "/market/ticker:all", expectedResult: []asset.Item{asset.Spot}},
		{pair: marginTradablePair, topic: "/market/ticker:all", expectedResult: []asset.Item{asset.Margin, asset.Spot}},
		{pair: futuresTradablePair, topic: "/contractMarket/tickerV2:XBTUSDM", expectedResult: []asset.Item{asset.Futures}},
	}
	for i := 0; i < b.N; i++ {
		for j := range cases {
			resp, err := ku.CalculateAssets(cases[j].topic, cases[j].pair)
			if err != nil {
				b.Fatal(err)
			}
			for x := range resp {
				var found bool
				for y := range cases[j].expectedResult {
					if resp[x] == cases[j].expectedResult[y] {
						found = true
						break
					}
				}
				if !found {
					b.Error("bad match")
				}
			}
		}
	}
}
```
here are the results:
```
BenchmarkButts-10     	   16918	     72570 ns/op	  268232 B/op	      25 allocs/op
BenchmarkButts2-10    	 1574653	       763.8 ns/op	      96 B/op	      13 allocs/op
```

File: exchanges/kucoin/kucoin_websocket.go [hunk 2]
```diff
@@ -75,8 +75,8 @@ const (
 )
 
 var subscriptionNames = map[string]string{
-	subscription.TickerChannel:    marketTickerChannel,
-	subscription.OrderbookChannel: marketOrderbookLevel2Channels,
+	subscription.TickerChannel:    marketAllTickersChannel,         // This allows more subscriptions on the orderbook channel for this specific connection.
```
Comment (gloriousCode) [inline exchanges/kucoin/kucoin_websocket.go:78]: A comment that I want to investigate further. If spot or margin is not enabled, this channel shouldn't be subscribed to, its wasteful given that it doesn't contain futures data.

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Bit of a funny review sorry 😆 I'll continue with it soon

File: exchanges/kucoin/kucoin_websocket.go (hunk 1 above)
Comment (gloriousCode) [reply exchanges/kucoin/kucoin_websocket.go:1727]: Updated to account for both private/public margin types and gone over each current use-case and kucoins docs.
Today's version is 6x faster than yesterday's and 420x (😎) faster than current. Hopefully it fits things better for you

File: exchanges/kucoin/kucoin_websocket.go (hunk 2 above)
Comment (gloriousCode) [reply exchanges/kucoin/kucoin_websocket.go:78]: In the function `expandSubscription` I think you can add the following check to it
```go
if len(assetPairs[s.Asset]) == 0 {
		return nil, nil
	}
```
That will prevent `ticker:all` being subscribed to if there are no spot pairs. I don't yet know whether ticker:all applies to margin pairs too though and may need additional changes

**Review by gloriousCode — APPROVED** [review]
tACK! 🎉

**Review by thrasher- — APPROVED** [review]
tACK, thanks for cleaning up the ghost errors too 😆

**Pull request merged**